service/history/workflow/activity.go

```diff
@@ -219,3 +220,149 @@ func PauseActivityById(mutableState MutableState, activityId string) error {
+	if err := mutableState.UpdateActivity(ai.ScheduledEventId, func(activityInfo *persistencepb.ActivityInfo, ms MutableState) {
+		activityInfo.Paused = false
+		activityInfo.Stamp++
+
+		// reset the number of attempts
+		activityInfo.Attempt = 1
+
+		if needRegenerateRetryTask(activityInfo, scheduleNewRun) {
+			scheduleTime := activityInfo.ScheduledTime.AsTime()
+			if scheduleNewRun {
+				scheduleTime = shardContext.GetTimeSource().Now().UTC()
+			}
+			_ = ms.RegenerateActivityRetryTask(activityInfo, scheduleTime)
+
+		}
+
+		if resetHeartbeats {
+			activityInfo.LastHeartbeatDetails = nil
+			activityInfo.LastHeartbeatUpdateTime = nil
+		}
+
+	}); err != nil {
+		return nil, err
+	}
+
+	return &historyservice.UnpauseActivityResponse{}, nil
+}
+
+func needRegenerateRetryTask(ai *persistencepb.ActivityInfo, scheduleNewRun bool) bool {
+	// if activity is Paused - we don't need to do anything
+	if ai.Paused {
+		return false
+	}
+
+	// we need to update the Stamp:
+	// * if scheduleNewRun flag is provided and activity is in retry
+	//		- we need to prevent the retry from executing
+	// * if scheduleNewRun flag is provided and activity is running
+	//		- we need to prevent the previous run from succeeding
+	// * if scheduleNewRun flag is not provided  and activity is in retry
+	//		- retry may not happen  since the number of attempts changes
+	//   	- because of that we need to generate new retry task
+	//  	- because of that we need to increase the Stamp to block previous retry attempt (if the number of attempts doesn't change)
+	// the only case when we don't need to increase the Stamp is when activity is running and there is no scheduleNewRun flag.
+
+	if scheduleNewRun && GetActivityState(ai) == enumspb.PENDING_ACTIVITY_STATE_SCHEDULED {
+		// activity is waiting to start, scheduleNewRun flag is provided
+		// we need to start the activity immediately
+		// and increase the Stamp to prevent old retry from executing
+		return true
+	}
+
+	if scheduleNewRun && GetActivityState(ai) != enumspb.PENDING_ACTIVITY_STATE_SCHEDULED {
+		// activity is running, scheduleNewRun flag is provided
+		// we need to start the activity immediately
+		// need to increase the Stamp to prevent old instance from succeeding
+		return true
+	}
+
+	if !scheduleNewRun && GetActivityState(ai) == enumspb.PENDING_ACTIVITY_STATE_SCHEDULED {
+		// activity is waiting to start, scheduleNewRun flag is NOT provided
+		// in this case we still need to increase the Stamp
+		// this is because we reset the number of attempts, and this will prevent old retry from executing
+		// we also need to start the activity immediately - no reason to wait
+		// we still need to increase the stamp in case the number of attempts doesn't change
+		return true
+	}
+
+	if !scheduleNewRun && GetActivityState(ai) != enumspb.PENDING_ACTIVITY_STATE_SCHEDULED {
+		// activity is running, scheduleNewRun flag is NOT provided
+		// in this case we don't need to do anything
```

Comment:
this is because there's no Stamp check when recording activity completed/failed?

Reply:
This is because in this case there is no need to generate new retry task:
1. Activity is running, so it is not in retry.
2. Activity is supposed to finish, because scheduleNewRun flag is not provided.